### DISCUSSION TITLE
soil moisture  sensor in calculation

### DISCUSSION THREAD

**Discussion by sat-fan2019** [altmenorg/HAsmartirrigation#385] [Ideas]
Hello. Thanks for this project! I have a separate soil moisture sensor for each zone. 
Is it possible to include this information in the calculation? I think that would be a much more accurate sensor than the "open Weather" vlaue humidity.  Especially if you have a plant house that has a completely different climate than outside. Thank you if you can make this possible!

**Comment by jeroenterheerdt** [comment]
From the readme: "Note that the humidity sensor mentioned is the air humidity / atmospheric humidity, not soil humidity."

**Reply by sat-fan2019** [reply]
yes ... I know ... but as I mentioned I have like a lot of people that I know something like that:
" Capacitive Soil Moisture Sensor Module Wide Voltage 3.3~5.5V With Cable Wire for Arduino"
and it would be a great feature to use these values. 
Regards Stefan

**Reply by jeroenterheerdt** [reply]
I have the same, but... what is measures is just not the same. Try it: water the soil and you'll see your moisture sensor spike and then drop. That is not how you measure how much water to add to the soil to make plants thrive. You simply don't look at soil moisture for that _at all_.

**Reply by evanfoster** [reply]
There are soil moisture sensors that are genuinely useful for this purpose, but they're expensive and are often quite large. They rely on something like what an ethernet cable length tester uses, called TDT (Time Domain Transmissometry. Cable testers use TDR). They shoot EM waves into the soil and measure the transmission time and phase and other things I don't understand in order to figure out the true soil humidity. 

Capacitive sensors like the one you mentioned are greatly affected by soil chemistry and don't really effectively measure soil humidity in an agricultural context. They're useful for telling you when a house plant might need water if you keep the soil chemistry relatively constant, but they're just not helpful here.

The cheapest sensor you can get that's not affected too much by soil chemistry is probably the [SMT100](https://www.truebner.de/en/smt100.php). It works in a fashion similar to a TDT sensor, but it's much smaller and is somewhat cheaper. Cheaper is relative, since the best price I found online was €129. One nice thing is that it's officially supported in ESPHome.

I'm not asking for this to be integrated, but I do think it's interesting to consider how a proper soil humidity sensor might be used. Like, I could see it being useful for catching irrigation issues (e.g. a plant has grown up around a sprinkler and is preventing it from effectively distributing water). You could have an automation or a separate integration that reads/calculates the estimated soil humidity from HAsmartirrigation and then compares it to sensors in various zones so you could be quickly alerted. A good sensor might also be a nice calibration input.

**Comment by malalaska** [comment]
Servus,
I also like the idea of measuring the soil moisture. I also use 3 soil moisture sensors (zigbee2mqtt). I have set approximate irrigation times depending on whether the soil moisture is less than 65%, 55% or less than 45%. 

It is currently raining a lot in GER. It is very warm and my sensor readings are around 90% soil moisture. After HAsmartirrigation calculation I had to irrigate full time, but it makes no sense

Perhaps you could also integrate something like this as an additional function.

**Comment by jeroenterheerdt** [comment]
I can't and I won't. It is simply not the scope of these kind of calculations that this integrations. It's about moisture loss and making up for it. Moisture sensors simply don't measure that and are too finicky for anything useful for this scenario. Put such a sensor on your grass beneath your sprinkler head. The moisture reading spikes immediately. So if we used that to decide to irrigate or not we would stop immediately after start. Useless.

Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by fcastilloec** [comment]
Where I live, weather data is completely unreliable, I'm in a rural area. I would like to use the moisture sensors to tell whether is raining or not, not to calculate the amount of water needed based on moisture. The sensor data can be a complement to the weather data, and treated as such. If weather data says it's raining, check with the outdoor moisture sensor for confirmation. It's also easy to disregard the moisture sensor measurements if the sprinklers are on, we know the "rain" is coming from irrigation.
A weather station with a rain gauge could also be used, but those are usually a lot more expensive than a couple of moisture sensors, and can't be use with this integration either.
It's clear that moisture sensors and what they actually measure is not good enough to make calculation on what and when to irrigate, but they can be great to confirm unreliable weather data.

**Comment by DetrimentL** [comment]
I installed 7 zigbee soil moisture sensors around the zones (1 per zone) and have a daily routine that modifies the bucket values prior to the scheduled SI calculations (when enabled). Actually, Claude did it. Here is what he says about implementation:

Soil Sensor → Smart Irrigation Bucket Sync
What It Does
Smart Irrigation (SI) uses a "bucket" value per zone to track the cumulative water deficit. Normally SI calculates this purely from weather data (ET, rainfall). This routine intercepts the bucket before each nightly calculation and adjusts it based on actual soil moisture readings from physical sensors, so SI's runtime recommendations reflect what the soil is actually holding rather than what the weather model predicts.

How It Works
A soil moisture reading (0–100%) is mapped to a bucket value using linear interpolation between two calibration points:

Field capacity (e.g. 35%) → bucket = 0 (soil is full, no watering needed)
Wilting point (e.g. 15%) → bucket = −max_bucket (soil is empty, maximum deficit)
Any reading above field capacity clamps to 0. Any reading below wilting point clamps to −max_bucket. The resulting bucket value is written directly to the SI zone via the smart_irrigation.set_bucket service call.

Timing
The sync runs at 22:50 local time, triggered by a time automation. This is intentionally 10 minutes before SI's nightly ET calculation at 23:00. SI reads the bucket at calculation time, so writing at 22:50 ensures the soil-adjusted value is what SI sees when it determines tonight's runtime. If the sync ran after 23:00 it would affect the next night's calculation instead.

Zone Mapping
Each SI zone maps to a specific soil sensor. In this implementation multiple zones can share a sensor (e.g. both lower yard zones use the same probe since they're adjacent). The mapping is defined in the automation variables:

zones:
  - si_entity: sensor.smart_irrigation_lower_yard_west
    soil_entity: sensor.my_weather_station_soil_humidity_1
    max_bucket: 24
  - si_entity: sensor.smart_irrigation_lower_yard_east
    soil_entity: sensor.my_weather_station_soil_humidity_1
    max_bucket: 24
  # ... one entry per zone
Calibration Inputs
Two input_number helpers control the interpolation thresholds and are adjustable from the dashboard without editing YAML:

input_number.irrigation_soil_threshold — wilting point % (default 15%). Below this, bucket is set to full deficit.
input_number.soil_field_capacity_percent — field capacity % (default 35%). Above this, bucket is set to 0.
The max_bucket per zone comes from the SI zone configuration itself.

The Math
if soil >= field_capacity:
    bucket = 0
elif soil <= wilting_point:
    bucket = -max_bucket
else:
    ratio = (soil - wilting_point) / (field_capacity - wilting_point)
    bucket = -max_bucket * (1 - ratio)
So at 25% moisture with field capacity 35% and wilting point 15%:

ratio = (25−15) / (35−15) = 0.5
bucket = −max_bucket × 0.5 → half deficit
What to Know Before Implementing
Sensor units must be 0–100%. The interpolation assumes a percentage. If your sensor outputs volumetric water content (m³/m³) or raw counts you'll need a template sensor to convert it first.

Each zone needs its own max_bucket value. This is found in the SI zone attributes (maximum_bucket). It varies by zone based on soil type, root depth, and other SI parameters you've configured. Using the wrong max_bucket will cause over- or under-watering.

Field capacity and wilting point are per-soil-type. The defaults (35% / 15%) are reasonable starting points for loam. Sandy soils have lower field capacity (~20–25%), clay soils higher (~40–50%). Calibrate by observing your sensor readings immediately after a thorough watering (field capacity) and just before plants show stress (wilting point).

This completely overrides SI's weather-based bucket accumulation. Once you enable the sync, SI's ET calculations still run but the bucket is reset from soil sensors every night, so multi-day ET accumulation is discarded. If your sensors are reliable this is more accurate; if a sensor goes unavailable or reads incorrectly, SI will water based on bad data. The automation includes an unavailability check and skips the sync for any zone whose sensor is unavailable, leaving the existing bucket value intact for that zone.

The sync only runs if input_boolean.irrigation_soil_sync_enabled is on. There's a dashboard toggle to disable the entire sync without modifying the automation, useful for troubleshooting or if you want SI to run purely on weather data temporarily.

Zones without a soil sensor are unaffected. SI manages their buckets normally via ET.

Dependencies
Smart Irrigation HACS integration with smart_irrigation.set_bucket service available
Physical soil moisture sensors integrated into HA (this setup uses Ecowitt/Fine Offset weather station probes and Bluetooth soil sensors)
input_number.irrigation_soil_threshold and input_number.soil_field_capacity_percent helpers
input_boolean.irrigation_soil_sync_enabled toggle helper

**Comment by altmenorg** [comment]
@sat-fan2019 @fcastilloec @DetrimentL a soil moisture sensor can now be used in [v2026.9.1-beta3](https://github.com/altmenorg/HAsmartirrigation/releases/tag/v2026.9.1-beta3), though not in the way this discussion first asked.

The original answer still holds for the calculation: a moisture reading does not measure how much water the soil lost, so it does not enter the water balance. What it can say reliably is that the soil is already wet. So each zone can take a soil moisture sensor and a threshold in %, and while the reading is at or above it, that zone sits the run out and the others water. Its bucket is kept, so the deficit rolls over to the next run. A sensor that cannot be read never stops a run.

For the case @fcastilloec describes, where the weather data cannot be trusted to say whether it rains, the same beta can also skip a run on a rain sensor, and on freeze or wind. The settings are under General, and the soil moisture on each zone.

Feedback from anyone running it on real sensors is very welcome here.
